File: examples/android-video-demo/app/src/main/java/com/giso/demo/video/tracking/PlaybackTracker.java
package com.giso.demo.video.tracking;

import android.os.Handler;
import android.os.Looper;

import com.giso.demo.video.model.VideoEpisode;
import com.giso.tracker.BizEvents;
import com.giso.tracker.Params;
import com.giso.tracker.Tracker;

import java.util.HashMap;
import java.util.Map;

/**
 * 播放业务事件：video_play_start / heartbeat / end。
 * 对应 schema 登记的长视频领域事件，演示 biz_event 上报。
 */
public final class PlaybackTracker {
    private static final long HEARTBEAT_MS = 30_000L;

    private final Tracker tracker = Tracker.get();
    private final Handler handler = new Handler(Looper.getMainLooper());
    private final Runnable heartbeatTask = this::emitHeartbeat;

    private VideoEpisode episode;
    private long playStartMs;
    private long accumulatedMs;
    private long lastTickMs;
    private boolean started;
    private float speed = 1.0f;

    public void onEpisodeReady(VideoEpisode ep, boolean autoPlay) {
        if (started) return;
        episode = ep;
        started = true;
        playStartMs = System.currentTimeMillis();
        lastTickMs = playStartMs;
        Map<String, Object> p = baseParams();
        p.put(Params.IS_AUTO, autoPlay ? 1 : 0);
        tracker.bizEvent(BizEvents.VIDEO_PLAY_START, p);
        scheduleHeartbeat();
    }

    public void onPlaying(long positionMs) {
        lastTickMs = System.currentTimeMillis();
    }

    public void onPaused(long positionMs) {
        tick(positionMs);
        handler.removeCallbacks(heartbeatTask);
        emitHeartbeat(positionMs);
    }

    public void onSpeedChanged(float speed) {
        this.speed = speed;
    }

    public void onEnded(long positionMs) {
        finish(positionMs, true);
    }

    public void onStopped(long positionMs) {
        finish(positionMs, false);
    }

    private void finish(long positionMs, boolean completed) {
        if (!started) return;
        tick(positionMs);
        handler.removeCallbacks(heartbeatTask);
        Map<String, Object> p = baseParams();
        p.put(Params.PLAY_DUR, accumulatedMs);
        p.put(Params.PLAY_POS, positionMs);
        p.put(Params.VIDEO_DUR, episode.durationMs);
        tracker.bizEvent(BizEvents.VIDEO_PLAY_END, p);
        started = false;
    }

    public void onError(String reason) {
        if (episode == null) return;
        Map<String, Object> p = new HashMap<>();
        p.put(Params.VID, episode.vid);
        p.put(Params.FAIL_REASON, reason);
        tracker.bizEvent(BizEvents.VIDEO_PLAY_ERROR, p);
        started = false;
        handler.removeCallbacks(heartbeatTask);
    }

    private void emitHeartbeat() {
        if (!started || episode == null) return;
        emitHeartbeat(episode.durationMs > 0 ? accumulatedMs : 0);
        scheduleHeartbeat();
    }

    private void emitHeartbeat(long positionMs) {
        Map<String, Object> p = baseParams();
        p.put(Params.PLAY_DUR, accumulatedMs);
        p.put(Params.PLAY_POS, positionMs);
        p.put(Params.SPEED, (double) speed);
        tracker.bizEvent(BizEvents.VIDEO_PLAY_HEARTBEAT, p);
    }

    private void scheduleHeartbeat() {
        handler.removeCallbacks(heartbeatTask);
        handler.postDelayed(heartbeatTask, HEARTBEAT_MS);
    }

    private void tick(long positionMs) {
        long now = System.currentTimeMillis();
        if (lastTickMs > 0) {
            accumulatedMs += Math.max(0, now - lastTickMs);
        }
        lastTickMs = now;
    }

    private Map<String, Object> baseParams() {
        Map<String, Object> p = new HashMap<>();
        p.put(Params.VID, episode.vid);
        String series = episode.seriesId != null && !episode.seriesId.isEmpty()
                ? episode.seriesId : episode.vid;
        p.put(Params.SERIES_ID, series);
        p.put(Params.EP_NUM, episode.epNum > 0 ? episode.epNum : 1);
        p.put(Params.DEFINITION, episode.definition);
        return p;
    }
}

File: examples/flutter-video-demo/lib/tracking/playback_tracker.dart
import 'dart:async';

import 'package:giso_tracker/giso_tracker.dart';

import '../model/video_episode.dart';

/// Mirrors Android demo: video_play_start / heartbeat / end / error.
class PlaybackTracker {
  static const _heartbeatMs = Duration(seconds: 30);

  VideoEpisode? _episode;
  bool _started = false;
  int _accumulatedMs = 0;
  int _lastTickMs = 0;
  double _speed = 1.0;
  Timer? _heartbeatTimer;

  void onEpisodeReady(VideoEpisode ep, {required bool autoPlay}) {
    if (_started) return;
    _episode = ep;
    _started = true;
    _lastTickMs = DateTime.now().millisecondsSinceEpoch;
    final params = _baseParams();
    params[Params.isAuto] = autoPlay ? 1 : 0;
    GisoTracker.instance.bizEvent(BizEvents.videoPlayStart, params);
    _scheduleHeartbeat();
  }

  void onPlaying() {
    _lastTickMs = DateTime.now().millisecondsSinceEpoch;
  }

  void onPaused(int positionMs) {
    _tick(positionMs);
    _heartbeatTimer?.cancel();
    _emitHeartbeat(positionMs);
  }

  void onSpeedChanged(double speed) => _speed = speed;

  void onEnded(int positionMs) => _finish(positionMs);

  void onStopped(int positionMs) => _finish(positionMs);

  void onError(String reason) {
    final ep = _episode;
    if (ep == null) return;
    GisoTracker.instance.bizEvent(BizEvents.videoPlayError, {
      Params.vid: ep.vid,
      Params.failReason: reason,
    });
    _started = false;
    _heartbeatTimer?.cancel();
  }

  void dispose() {
    _heartbeatTimer?.cancel();
    if (_started) {
      _finish(_accumulatedMs);
    }
  }

  void _finish(int positionMs) {
    if (!_started || _episode == null) return;
    _tick(positionMs);
    _heartbeatTimer?.cancel();
    final params = _baseParams();
    params[Params.playDur] = _accumulatedMs;
    params[Params.playPos] = positionMs;
    params[Params.videoDur] = _episode!.durationMs;
    GisoTracker.instance.bizEvent(BizEvents.videoPlayEnd, params);
    _started = false;
  }

  void _emitHeartbeat([int? positionMs]) {
    final ep = _episode;
    if (!_started || ep == null) return;
    final pos = positionMs ?? (_accumulatedMs > 0 ? _accumulatedMs : 0);
    final params = _baseParams();
    params[Params.playDur] = _accumulatedMs;
    params[Params.playPos] = pos;
    params[Params.speed] = _speed;
    GisoTracker.instance.bizEvent(BizEvents.videoPlayHeartbeat, params);
  }

  void _scheduleHeartbeat() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = Timer.periodic(_heartbeatMs, (_) => _emitHeartbeat());
  }

  void _tick(int positionMs) {
    final now = DateTime.now().millisecondsSinceEpoch;
    if (_lastTickMs > 0) {
      _accumulatedMs += (now - _lastTickMs).clamp(0, 1 << 30);
    }
    _lastTickMs = now;
  }

  Map<String, Object?> _baseParams() {
    final ep = _episode!;
    final series = ep.seriesId.isNotEmpty ? ep.seriesId : ep.vid;
    return {
      Params.vid: ep.vid,
      Params.seriesId: series,
      Params.epNum: ep.epNum > 0 ? ep.epNum : 1,
      Params.definition: ep.definition,
    };
  }
}
